File: code/frozen_lake/population.py
"""population
population class for genetic algorithm. Contains initiation functions,
calls for fitness fvalues, creates genepool and regulates reproduction.
"""

from typing import List
import numpy as np

from individual import Individual


class Population():
    """
    Control population during running the genetic algorithm. Has
    functions for creating a population, producing offspring and
    scoring the population.

    Parameters:
        genes (str): options to create dna from
        population_size (int): amount of individuals per generation
        target (str): target of evolution
        target_length (int): length of target
        mutation_rate (float): determines mutation during reproduction
    """

    def __init__(self,
                 env,
                 population_size: int,
                 mutation_rate: float,
                 dna_length: int):


        self.env = env

        self.population: List[Individual] = []
        self.population_size = population_size

        self.mutation_rate = mutation_rate
        self.dna_length = dna_length

        self.max_fitness: float = 0.0

    def init_population(self):
        """
        Create n individuals where n is the population
        size. Save these in self.population.
        """

        for _ in range(self.population_size):
            self.population.append(Individual(
                self.env,
                self.dna_length))


    def _create_p_list(self):
        """
        Create a list of fitness scores and scale it to have a sum
        of 1.
        """

        p_list = np.array([ind.fitness for ind in self.population])
        return p_list / np.sum(p_list)


    def create_offspring(self):
        """
        For population size, pick two individuals from the population
        based on the fitness score and create offspring, this will be
        the new population.
        """

        new_population = []
        p_list = self._create_p_list()
        for _ in range(self.population_size):
            # Select to individuals, chance is scaled to fitness score
            parents = np.random.choice(self.population, 2, p=p_list)
            # Create new dna from parents
            new_dna = parents[0].mix_dna(
                parents[1].dna, self.mutation_rate)
            new_population.append(Individual(
                self.env,
                self.dna_length,
                new_dna))
        self.population = new_population


    def calc_fitness(self):
        """
        Sort the population and extract the maximal fitness from
        the population
        """
        self.population = sorted(
            self.population,
            key=lambda x: x.fitness,
            reverse=True)
        self.max_fitness = self.population[0].fitness

File: presentation/to_be/population.js
// Container for population
class Population {
  constructor(pop) {
    this.pop = pop;
    this.generation = 0;
  }

  // Create initial population
  initPopulation() {
    let init = [];
    for (let i = 0; i < populationSize; i++) {
      let ind = new Individual();
      ind.initGenome();
      ind.calcFitness();
      init.push(ind);
    }
    this.pop = init;
  }

  // Create population
  createPool() {
    this.pool = [];
    for (let i = 0; i < this.pop.length; i++) {
      let n = Math.floor(this.pop[i].fitness * 100);
      for (let j = 0; j < n; j++) {
        this.pool.push(this.pop[i]);
      }
    }
  }

  // Create offsprint
  offspring(a, b) {
    let splitVal = Math.floor(Math.random() * a.genome.length);
    let newGenome =
      a.genome.slice(0, splitVal) + b.genome.slice(splitVal, a.genome.length);
    newGenome = newGenome.split("");
    for (let i = 0; i < newGenome.length; i++) {
      if (Math.random() < mutationRate) {
        newGenome[i] = genes[Math.floor(Math.random() * genes.length)];
      }
    }
    return newGenome.join``;
  }

  // Create new population
  newPopulation() {
    let p = [];
    for (let i = 0; i < populationSize; i++) {
      let a = this.pool[Math.floor(Math.random() * this.pool.length)];
      let b = this.pool[Math.floor(Math.random() * this.pool.length)];
      let ind = new Individual(this.offspring(a, b));
      ind.calcFitness();
      p.push(ind);
    }
    this.pop = p;
  }

  // Calculate max fitnes of population
  sortFitness() {
    this.pop.sort((a, b) => {
      return b.fitness - a.fitness;
    });
  }
}
